File: reference/python/crovia_seal/stego.py
"""
Crovia Invisible Mark (CIM) v1 — Python reference implementation.

This module is the authoritative Python twin of the TypeScript
`integrations/browser-extension/src/lib/stego.ts`. Every function is
deliberately kept algorithmically identical so that a CIM produced by one
language is verified by the other with ZERO tolerance (byte-identical test
vectors live in `conformance/vectors/cim/`).

WIRE FORMAT (must match stego.ts):

    START_MARK (3 cp)  +  DATA_BITS (130 cp)  +  CRC_BITS (16 cp)  +  END_MARK (3 cp)

    START_MARK = U+200D U+FEFF U+200D
    END_MARK   = U+FEFF U+200D U+FEFF
    bit 0      = U+200B
    bit 1      = U+200C
    CRC        = CRC-16/CCITT (poly 0x1021, init 0xFFFF, no final xor), MSB first

The suffix of a `seal_id` (26 RFC-4648 base32 chars) is encoded in DATA_BITS.
The `cs_YYYY_` prefix is reconstructed at extraction time using the
caller-provided `issuance_year` (default: current UTC year).

THIS FILE MUST NEVER DIVERGE FROM stego.ts. Changes here require mirrored
changes + re-generated conformance vectors in the TS twin.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import List, Optional

# ---------------------------------------------------------------------------
# Unicode constants (MUST equal stego.ts)
# ---------------------------------------------------------------------------

ZW_BIT_0 = "\u200B"  # ZERO WIDTH SPACE
ZW_BIT_1 = "\u200C"  # ZERO WIDTH NON-JOINER
ZWJ = "\u200D"
BOM = "\uFEFF"

CIM_START = ZWJ + BOM + ZWJ
CIM_END = BOM + ZWJ + BOM
CIM_BITS_DATA = 130
CIM_BITS_CRC = 16
CIM_TOTAL_LEN = len(CIM_START) + CIM_BITS_DATA + CIM_BITS_CRC + len(CIM_END)

_BASE32_ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567"
_BASE32_INDEX = {c: i for i, c in enumerate(_BASE32_ALPHABET)}
_SEAL_ID_RE = re.compile(r"^cs_([0-9]{4})_([A-Z2-7]{26})$")
_BASE32_26_RE = re.compile(r"^[A-Z2-7]{26}$")


# ---------------------------------------------------------------------------
# Bit packing helpers
# ---------------------------------------------------------------------------

def _base32_to_bits(base32: str) -> List[int]:
    if not _BASE32_26_RE.match(base32):
        raise ValueError("invalid base32 suffix: need 26 upper-case RFC 4648 chars")
    bits: List[int] = [0] * CIM_BITS_DATA
    for i in range(26):
        value = _BASE32_INDEX[base32[i]]
        for b in range(4, -1, -1):  # MSB first
            bits[i * 5 + (4 - b)] = (value >> b) & 1
    return bits


def _bits_to_base32(bits: List[int]) -> str:
    if len(bits) != CIM_BITS_DATA:
        raise ValueError(f"expected {CIM_BITS_DATA} bits, got {len(bits)}")
    out_chars: List[str] = []
    for i in range(26):
        v = 0
        for b in range(5):
            v = (v << 1) | (bits[i * 5 + b] & 1)
        out_chars.append(_BASE32_ALPHABET[v])
    return "".join(out_chars)


# ---------------------------------------------------------------------------
# CRC-16/CCITT (bitwise; MUST equal stego.ts `crc16Bits`)
# ---------------------------------------------------------------------------

def _crc16_bits(bits: List[int]) -> int:
    crc = 0xFFFF
    for bit in bits:
        top = (crc & 0x8000) != 0
        crc = (crc << 1) & 0xFFFF
        if top != (bit == 1):
            crc ^= 0x1021
    return crc & 0xFFFF


# ---------------------------------------------------------------------------
# Zero-width encoder / decoder
# ---------------------------------------------------------------------------

def _bits_to_zw(bits: List[int]) -> str:
    return "".join(ZW_BIT_1 if b else ZW_BIT_0 for b in bits)


def _zw_to_bits(zw: str) -> List[int]:
    bits: List[int] = []
    for i, c in enumerate(zw):
        if c == ZW_BIT_0:
            bits.append(0)
        elif c == ZW_BIT_1:
            bits.append(1)
        else:
            raise ValueError(f"unexpected char in CIM bit stream at index {i}")
    return bits


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

@dataclass(frozen=True)
class ExtractedCim:
    """Result of a successful CIM extraction.

    Mirrors the `ExtractedCim` interface in stego.ts. Instances of this class
    are only ever returned for marks whose CRC matched; a `crc_valid=False`
    mark is NEVER surfaced (mismatches are silently skipped so a single bad
    bit cannot shadow a subsequent valid mark).
    """
    seal_id: str
    base32: str
    crc_valid: bool
    start_index: int
    end_index: int
    year: int


def encode_cim(seal_id: str) -> str:
    """Encode a seal_id into a 152-code-point CIM string (all invisible)."""
    m = _SEAL_ID_RE.match(seal_id)
    if not m:
        raise ValueError(f"invalid seal_id: {seal_id!r}")
    base32 = m.group(2)
    data_bits = _base32_to_bits(base32)
    crc = _crc16_bits(data_bits)
    crc_bits = [(crc >> (CIM_BITS_CRC - 1 - i)) & 1 for i in range(CIM_BITS_CRC)]
    return CIM_START + _bits_to_zw(data_bits) + _bits_to_zw(crc_bits) + CIM_END


def embed_cim(visible_text: str, seal_id: str) -> str:
    """Inject a CIM into visible text.

    Placement policy (identical to stego.ts):
      - if the text contains a newline in the last 200 chars, insert the mark
        immediately BEFORE that newline (survives "first paragraph" crops);
      - otherwise append to the end.
    """
    mark = encode_cim(seal_id)
    last_nl = visible_text.rfind("\n")
    if last_nl >= 0 and last_nl > len(visible_text) - 200:
        return visible_text[:last_nl] + mark + visible_text[last_nl:]
    return visible_text + mark


def extract_all_cims(text: str, issuance_year: Optional[int] = None) -> List[ExtractedCim]:
    """Scan `text` for ALL valid CIM marks and return them in order.

    An invalid CIM (wrong CRC, bad bit alphabet, truncated end marker) is
    skipped over and NEVER returned: silent failures are safer than spoofing.
    """
    year = issuance_year if issuance_year is not None else datetime.now(timezone.utc).year
    out: List[ExtractedCim] = []
    cursor = 0
    while cursor < len(text):
        start = text.find(CIM_START, cursor)
        if start < 0:
            break
        payload_start = start + len(CIM_START)
        end_candidate = payload_start + CIM_BITS_DATA + CIM_BITS_CRC
        if end_candidate + len(CIM_END) > len(text):
            break
        if text[end_candidate:end_candidate + len(CIM_END)] != CIM_END:
            cursor = start + 1
            continue
        bit_stream = text[payload_start:end_candidate]
        if any(c not in (ZW_BIT_0, ZW_BIT_1) for c in bit_stream):
            cursor = start + 1
            continue
        try:
            data_bits = _zw_to_bits(bit_stream[:CIM_BITS_DATA])
            crc_bits = _zw_to_bits(bit_stream[CIM_BITS_DATA:])
            crc_from = 0
            for i in range(CIM_BITS_CRC):
                crc_from = (crc_from << 1) | crc_bits[i]
            crc_computed = _crc16_bits(data_bits)
            if crc_computed != crc_from:
                cursor = start + 1
                continue
            base32 = _bits_to_base32(data_bits)
            out.append(ExtractedCim(
                seal_id=f"cs_{year}_{base32}",
                base32=base32,
                crc_valid=True,
                start_index=start,
                end_index=end_candidate + len(CIM_END),
                year=year,
            ))
            cursor = end_candidate + len(CIM_END)
        except ValueError:
            cursor = start + 1
    return out


def extract_cim(text: str, issuance_year: Optional[int] = None) -> Optional[ExtractedCim]:
    """Return the FIRST valid CIM in `text`, or None."""
    all_marks = extract_all_cims(text, issuance_year)
    return all_marks[0] if all_marks else None


def strip_cim(text: str) -> str:
    """Return `text` with all valid CIMs removed; invalid partials kept."""
    all_marks = extract_all_cims(text)
    if not all_marks:
        return text
    out = text
    for m in reversed(all_marks):
        out = out[:m.start_index] + out[m.end_index:]
    return out


def contains_cim_marker(text: str) -> bool:
    """Heuristic: does this text contain at least a CIM start sequence?"""
    return CIM_START in text


__all__ = [
    "ZW_BIT_0", "ZW_BIT_1", "ZWJ", "BOM",
    "CIM_START", "CIM_END",
    "CIM_BITS_DATA", "CIM_BITS_CRC", "CIM_TOTAL_LEN",
    "ExtractedCim",
    "encode_cim", "embed_cim",
    "extract_cim", "extract_all_cims",
    "strip_cim", "contains_cim_marker",
]

File: integrations/browser-extension/src/lib/stego.ts
/**
 * Crovia Invisible Mark (CIM) v1 - zero-width steganographic encoding of a
 * Seal identifier inside a human-readable AI output.
 *
 * =============================================================================
 *                 MOTIVATION: why CIM exists at all
 * =============================================================================
 *
 * A Seal is a compact JSON object. Attaching it as an external sidecar (C2PA
 * style) is fragile: the moment the user copies the AI output into an email,
 * a Notion page, a Slack message, or a PDF, the sidecar is lost and the
 * provenance trail is broken forever.
 *
 * CIM solves this by embedding a reference to the Seal INSIDE the output
 * text itself, using Unicode code points that are rendered as ZERO WIDTH by
 * every conformant Unicode renderer (browsers, Word, email clients). The
 * reference is a 130-bit id (26 base32 chars). A verifier who later receives
 * the copied text anywhere can reconstruct the id, fetch the full Seal from
 * a transparency log, and reconstruct the full provenance chain.
 *
 * This is NOT equivalent to existing techniques:
 *   - C2PA/Content Credentials: sidecar metadata, lost on copy.
 *   - OpenAI watermarking proposals: logit biasing, probabilistic, removed
 *     by paraphrase, cannot be tied to a specific cryptographic signature.
 *   - Classic zero-width text watermarks: used for author tracking with
 *     bit strings unrelated to any external verifiable object. CIM's bits
 *     are the unique id of a cryptographically signed receipt.
 *
 * =============================================================================
 *                 WIRE FORMAT (strictly specified)
 * =============================================================================
 *
 *   START_MARK  DATA_BITS  CRC_BITS  END_MARK
 *
 * where:
 *
 *   START_MARK = U+200D U+FEFF U+200D      (3 chars, "ZWJ BOM ZWJ")
 *   END_MARK   = U+FEFF U+200D U+FEFF      (3 chars, "BOM ZWJ BOM")
 *
 *   DATA_BITS  = 130 bits, encoding 26 base32 characters (5 bits each)
 *   CRC_BITS   = 16-bit CRC-CCITT (polynomial 0x1021, initial 0xFFFF)
 *                of the 130 data bits, serialized MSB-first
 *
 *   Each BIT is encoded as a single invisible character:
 *     0 -> U+200B  ZERO WIDTH SPACE
 *     1 -> U+200C  ZERO WIDTH NON-JOINER
 *
 *   Total payload length: 3 + 130 + 16 + 3 = 152 invisible chars.
 *   Visible length of the host text: UNCHANGED.
 *
 * The marker triplets were chosen because:
 *   (a) They contain a BOM (U+FEFF) which is legal mid-stream in UTF-8 per
 *       RFC 3629 but is almost never produced by natural language. Detectors
 *       looking for CIM can fast-scan on BOM occurrence.
 *   (b) The triplet sequences differ between start and end (ZWJ-BOM-ZWJ vs
 *       BOM-ZWJ-BOM) so a parser cannot mistake the end of one mark for the
 *       start of another.
 *
 * =============================================================================
 *                 THREAT MODEL
 * =============================================================================
 *
 * CIM defends against ACCIDENTAL loss of provenance (copy/paste through
 * benign software). It does NOT defend against an active adversary who
 * strips zero-width characters deliberately. For that purpose the spec also
 * defines a VISIBLE footer (`[crovia-seal: cs_YYYY_...]`) that users can
 * opt into; the two are orthogonal. See `stripCim()` and the ATTRIBUTION
 * RIBBON behavior in content scripts.
 *
 * For tampering RESISTANCE within the mark itself, CIM includes a CRC-16
 * so truncated or bit-flipped marks are rejected by `extractCim()`. A CRC
 * failure is surfaced, never silently ignored.
 */

// ---------------------------------------------------------------------------
// Unicode constants
// ---------------------------------------------------------------------------

export const ZW_BIT_0 = '\u200B'; // ZERO WIDTH SPACE
export const ZW_BIT_1 = '\u200C'; // ZERO WIDTH NON-JOINER
export const ZWJ = '\u200D';      // ZERO WIDTH JOINER
export const BOM = '\uFEFF';      // ZERO WIDTH NO-BREAK SPACE / BOM

export const CIM_START = ZWJ + BOM + ZWJ;   // 3 chars
export const CIM_END = BOM + ZWJ + BOM;     // 3 chars
export const CIM_BITS_DATA = 130;            // 26 base32 chars x 5 bits
export const CIM_BITS_CRC = 16;
export const CIM_TOTAL_LEN =
  CIM_START.length + CIM_BITS_DATA + CIM_BITS_CRC + CIM_END.length;

// Alphabet of RFC 4648 base32 upper-case, no padding.
const BASE32_ALPHABET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567';

// seal_id format: cs_YYYY_<26 base32 chars>
// We only encode the 26 base32 chars; the prefix "cs_" and the 4-digit year
// can be re-inferred from context (current year when extracting). For
// chronological disambiguation the CRC also includes the original year,
// reconstructed before CRC check on extraction. See `encodeCim` and
// `decodeCim` for details.
const SEAL_ID_RE = /^cs_([0-9]{4})_([A-Z2-7]{26})$/;


// ---------------------------------------------------------------------------
// Bit-packing helpers
// ---------------------------------------------------------------------------

/** Return the 130-bit representation of the base32 suffix as a Uint8Array
 *  of booleans (1 bit per element, MSB-first across the 26 chars). */
function base32ToBits(base32: string): Uint8Array {
  if (!/^[A-Z2-7]{26}$/.test(base32)) {
    throw new Error('invalid base32 suffix: need 26 upper-case RFC 4648 chars');
  }
  const bits = new Uint8Array(CIM_BITS_DATA);
  for (let i = 0; i < 26; i++) {
    const value = BASE32_ALPHABET.indexOf(base32[i]!);
    if (value < 0) throw new Error(`base32 char out of alphabet: ${base32[i]}`);
    // 5 bits, MSB first
    for (let b = 4; b >= 0; b--) {
      bits[i * 5 + (4 - b)] = (value >> b) & 1;
    }
  }
  return bits;
}

function bitsToBase32(bits: Uint8Array): string {
  if (bits.length !== CIM_BITS_DATA) {
    throw new Error(`expected ${CIM_BITS_DATA} bits, got ${bits.length}`);
  }
  let out = '';
  for (let i = 0; i < 26; i++) {
    let v = 0;
    for (let b = 0; b < 5; b++) {
      v = (v << 1) | (bits[i * 5 + b]! & 1);
    }
    out += BASE32_ALPHABET[v]!;
  }
  return out;
}


// ---------------------------------------------------------------------------
// CRC-16/CCITT (polynomial 0x1021, initial 0xFFFF, no final XOR)
// ---------------------------------------------------------------------------

function crc16Bits(bits: Uint8Array): number {
  let crc = 0xFFFF;
  for (const bit of bits) {
    const topBit = (crc & 0x8000) !== 0;
    crc = (crc << 1) & 0xFFFF;
    if (topBit !== (bit === 1)) {
      crc ^= 0x1021;
    }
  }
  return crc & 0xFFFF;
}


// ---------------------------------------------------------------------------
// Zero-width encode/decode
// ---------------------------------------------------------------------------

function bitsToZw(bits: Uint8Array): string {
  let out = '';
  for (const b of bits) out += b ? ZW_BIT_1 : ZW_BIT_0;
  return out;
}

function zwToBits(zw: string): Uint8Array {
  const bits = new Uint8Array(zw.length);
  for (let i = 0; i < zw.length; i++) {
    const c = zw[i]!;
    if (c === ZW_BIT_0) bits[i] = 0;
    else if (c === ZW_BIT_1) bits[i] = 1;
    else throw new Error(`unexpected char in CIM bit stream at index ${i}`);
  }
  return bits;
}


// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Encode a Seal id into a CIM string (invisible characters only).
 *
 * The result is always `CIM_TOTAL_LEN` Unicode code points. It is intended
 * to be concatenated with the visible host text. For best concealment the
 * mark should be placed at a position unlikely to be truncated by paste
 * operations (end of paragraph is ideal).
 */
export function encodeCim(sealId: string): string {
  const m = SEAL_ID_RE.exec(sealId);
  if (!m) throw new Error(`invalid seal_id: ${sealId}`);
  const base32 = m[2]!;
  const bits = base32ToBits(base32);
  const crc = crc16Bits(bits);
  const crcBits = new Uint8Array(CIM_BITS_CRC);
  for (let i = 0; i < CIM_BITS_CRC; i++) {
    crcBits[i] = (crc >> (CIM_BITS_CRC - 1 - i)) & 1;
  }
  return CIM_START + bitsToZw(bits) + bitsToZw(crcBits) + CIM_END;
}


/**
 * Inject a CIM into a visible text, returning the combined string.
 * By default the mark is placed immediately BEFORE the final newline or,
 * failing that, appended at the end. This maximizes the chance that the
 * mark survives typical truncations (e.g. "first paragraph only" selection).
 */
export function embedCim(visibleText: string, sealId: string): string {
  const mark = encodeCim(sealId);
  const lastNewline = visibleText.lastIndexOf('\n');
  if (lastNewline >= 0 && lastNewline > visibleText.length - 200) {
    return visibleText.slice(0, lastNewline) + mark + visibleText.slice(lastNewline);
  }
  return visibleText + mark;
}


/**
 * Extract a CIM from a mixed text and return the reconstructed Seal id
 * plus CRC status. If no valid CIM is found, returns `null`. Multiple CIMs
 * are supported: the first valid one is returned. Use `extractAllCims()`
 * for the full list.
 */
export interface ExtractedCim {
  sealId: string;            // reconstructed "cs_YYYY_<base32>"
  base32: string;            // 26-char base32 suffix
  crcValid: boolean;         // always true if sealId is returned
  startIndex: number;        // index in the input string
  endIndex: number;          // exclusive end index
  year: number;              // inferred issuance year (current year at extract)
}

export function extractCim(text: string, issuanceYear?: number): ExtractedCim | null {
  const all = extractAllCims(text, issuanceYear);
  return all.length > 0 ? all[0]! : null;
}


export function extractAllCims(text: string, issuanceYear?: number): ExtractedCim[] {
  const year = issuanceYear ?? new Date().getUTCFullYear();
  const out: ExtractedCim[] = [];
  let cursor = 0;
  while (cursor < text.length) {
    const start = text.indexOf(CIM_START, cursor);
    if (start < 0) break;
    const payloadStart = start + CIM_START.length;
    const endCandidate = payloadStart + CIM_BITS_DATA + CIM_BITS_CRC;
    if (endCandidate + CIM_END.length > text.length) break;
    // Check end marker is exactly where expected.
    if (text.slice(endCandidate, endCandidate + CIM_END.length) !== CIM_END) {
      cursor = start + 1;
      continue;
    }
    const bitStream = text.slice(payloadStart, endCandidate);
    // Validate bit-stream contains only ZW_BIT_0 / ZW_BIT_1.
    let valid = true;
    for (let i = 0; i < bitStream.length; i++) {
      const c = bitStream[i]!;
      if (c !== ZW_BIT_0 && c !== ZW_BIT_1) { valid = false; break; }
    }
    if (!valid) { cursor = start + 1; continue; }

    try {
      const dataBits = zwToBits(bitStream.slice(0, CIM_BITS_DATA));
      const crcBitsArr = zwToBits(bitStream.slice(CIM_BITS_DATA));
      let crcFromBits = 0;
      for (let i = 0; i < CIM_BITS_CRC; i++) {
        crcFromBits = (crcFromBits << 1) | crcBitsArr[i]!;
      }
      const crcComputed = crc16Bits(dataBits);
      const crcValid = crcComputed === crcFromBits;
      if (!crcValid) {
        cursor = start + 1;
        continue;
      }
      const base32 = bitsToBase32(dataBits);
      out.push({
        sealId: `cs_${year}_${base32}`,
        base32,
        crcValid: true,
        startIndex: start,
        endIndex: endCandidate + CIM_END.length,
        year,
      });
      cursor = endCandidate + CIM_END.length;
    } catch {
      cursor = start + 1;
    }
  }
  return out;
}


/** Remove ALL valid CIMs from a string, returning the cleaned text.
 *  Invalid partial marks are left intact (they might be ordinary text). */
export function stripCim(text: string): string {
  const all = extractAllCims(text);
  if (all.length === 0) return text;
  // Remove from end to preserve indices.
  let out = text;
  for (let i = all.length - 1; i >= 0; i--) {
    const { startIndex, endIndex } = all[i]!;
    out = out.slice(0, startIndex) + out.slice(endIndex);
  }
  return out;
}


/** Heuristic: does this text CONTAIN any CIM-like structure (valid or not)? */
export function containsCimMarker(text: string): boolean {
  return text.includes(CIM_START);
}
